Numeric type literal syntax (#2015)
carbon-language/carbon-lang: proposals/p2015.md

```

Capture group 1 indicates either an ‘i’ for a two’s complement signed integer
type, a ‘u’ for an unsigned integer type, or a ‘f’ for a floating-point number
type. And capture group 2 specifies the width in bits.

Examples of this syntax include:

-   `i32` - A 32-bit two’s complement signed integer type
-   `u16` - A 16-bit unsigned integer type
-   `f128` - A 128-bit IEEE 754 floating-point number type

### Usage

```carbon
package sample api;

fn Sum(x: i32, y: i32) -> i32 {
  return x + y;
}

fn Main() -> i32 {
  return Sum(4, 2);
}
```

In the above example, `Sum` has parameters `x` and `y`, each of which is typed
as a 32-bit two’s complement signed integer. `Main` then returns the output of
`Sum` as a 32-bit two’s complement signed integer.

### Syntax examples and considerations

These examples are not meant to be formal mappings. They are only used to
illustrate to the reader how the numeric type syntax may connect to the
underlying language semantics in future numeric type additions.

#### Syntax example use in common fixed-size numeric types

[As proposed by @chandlerc](https://github.com/carbon-language/carbon-lang/issues/543#issuecomment-845620894)

> , our heuristic for type qualification here will be "Any truly pervasive

Reviewer: ```suggestion
[As proposed by @chandlerc](https://github.com/carbon-language/carbon-lang/issues/543#issuecomment-845620894),
our heuristic for type qualification here will be "Any truly pervasive
```

non-aggregate/composite/compound type that is also fundamentally
size-parameterized."

These fixed-width sizes can naturally cover the common register widths of 8-,
16-, 32-, 64-, and 128-bits, as seen in other languages.

Which expand to:

-   `i8`, `i16`, `i32`, `i64`, and `i128` keyword aliases for
    `Carbon.Int(N={8,16,32,64,128})`
-   `u8`, `u16`, `u32`, `u64`, and `u128` keyword aliases for
    `Carbon.UInt(N={8,16,32,64,128})`
-   `f16`, `f32`, `f64`, and `f128` keyword aliases for

>     `Carbon.Float(N={16,32,64,128})`

Reviewer: `f80` is another potentially interesting option, though it is currently only available on x86 platforms. See https://en.wikipedia.org/wiki/Extended_precision . Similarly, I'm not sure `f128` is available on all platforms.

Author: Good point. Similar to the previous comment, do you think we should add `f80` to the list of examples or just remove `f128`? Or perhaps move them to a platform-specific note.

The current list was pulled from the design doc's [primitive types section](https://github.com/carbon-language/carbon-lang/tree/trunk/docs/design#primitive-types).

Reviewer: That section wasn't really authoritative. I was revisiting it in #1975 . But really this proposal is the thing making a design in this space official. It should reflect the consensus of discussions and things like #543 and #1998 , but ultimately this proposal should stand on its own. This means this proposal can decide what the list is, but should also include a rationale for that choice. My suggestion is what I was going to put in #1975:

> Other sizes may be available, depending on the platform, such as [`f80`](https://en.wikipedia.org/wiki/Extended_precision), [`f128`](https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format), or [`f256`](https://en.wikipedia.org/wiki/Octuple-precision_floating-point_format).

The rationale is that Carbon should be exposing all the capabilities of the platform, in service of ["No need for a lower level language" in Carbon's "Performance-critical software" goal](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/goals.md#performance-critical-software).

Author: Yep, I'm in alignment there as well. I brought in those suggestions.

#### Syntax example use in arbitrary fixed-size numeric types

In the more general case, this syntax can express N-width types.

-   `iN` for `Carbon.Int(N)`, An N-width two’s complement signed integer type
-   `uN` for `Carbon.UInt(N)`, An N-width unsigned integer type

> -   `fN` for `Carbon.Float(N)`, An N-width IEEE 754 floating-point number type

Reviewer: I personally feel like we should limit the short syntax to the common sizes. The only use case I am aware of for things like `i3` are for [bit fields](https://en.wikipedia.org/wiki/Bit_field), and I feel like adding them to the language before we have a design for bit fields is premature. Furthermore, I think there are big benefits to limiting the syntax:

- The compiler can reject typos like `i23`, `i15`, etc.
- `i1`, `i2`, ... currently appear in C++ code in practice, examples:
    - https://github.com/google/googletest/blob/main/googlemock/include/gmock/gmock-matchers.h#L878
    - https://chromium.googlesource.com/external/github.com/abseil/abseil-cpp/+/HEAD/absl/container/btree_test.cc#2772
    - https://sourcegraph.com/search?q=context:global+lang:c%2B%2B+%5Ei1%24+type:symbol&patternType=regexp&case=yes
- Rules for other sizes are different, for example, we can't support pointers to arbitrary bits.
- Potential for confusion: `i1` and `il`, `i18` and `i18n`

I think this is a case where we should carefully consider the alternatives and have a clear rationale for the choice we make.

Author: I agree with this as well. As you stated, limiting to common sizes removes a bunch of visual confusion in the type names and only affects relatively fringe type cases. Do you think this should be reflected here in this proposal at all? Perhaps a note around this section? Or would it be more appropriate to remove it completely?

I feel like as this is a proposal centered around the syntax, it's not unreasonable to have this material covered here for the future, as long as it's clear to the reader.

Reviewer 2: I'd suggest starting with multiples of 8 for integer types, but allowing all multiples of 8 (other than 0). LLVM has really good support for this in most operations, it only struggles with things like division and a few other oddities. We should (IMO) consider adding limits when we need to rather than front-loading them here, and I think that restriction should be on the aliased thing (`Carbon.Int(102400)` or however its spelled), not on the literal syntax.

I'd actually suggest following the same pattern for floating point *literals* as well. While we'll immediately need to restrict the set of `Carbon.Float(N)` types to the values of `N` we can implement, the restriction should be there in the library and not in the literal syntax IMO.

## Rationale

Following Carbon’s goal to facilitate
["Code that is easy to read, understand, and write"](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write),
an explicit goal is to provide excellent ergonomics.

Highlighting relevant aspects of this from the project goals:

-   _Carbon should not use symbols that are difficult to type, see, or
    differentiate from similar symbols in commonly used contexts._
-   _Syntax should be easily parsed and scanned by any human in any development
    environment, not just a machine or a human aided by semantic hints from an
    IDE._
-   _Explicitness must be balanced against conciseness, as verbosity and
    ceremony add cognitive overhead for the reader, while explicitness reduces
    the amount of outside context the reader must have or assume._

## Alternatives considered

As discussed in
[#543](https://github.com/carbon-language/carbon-lang/issues/543), three other
options were considered:

> ### 1. C++ LP64 convention

Reviewer: I don't think the alternatives should be numbered. That doesn't match other proposals, and there is nothing associating these particular alternatives with these numbers.

## Background

For developer convenience, names are given to number types that map to native
machine register widths. These sizes typically include 8-bit, 16-bit, 32-bit,
64-bit, and, more recently, 128-bit widths.

For example, in [C++11+](https://en.cppreference.com/w/cpp/types/integer),
integer types such as `int8_t` (8-bit two's complement signed integer type) and
`uint16_t` (16-bit unsigned integer type) exist, among similar types for 32- and
64-bit values. Correspondingly, you have the `i8` and `u16`
([among others](https://doc.rust-lang.org/book/ch03-02-data-types.html#scalar-types))
scalar integer types in Rust. And in Swift, the `Int8` and `UInt16`
([among others](https://developer.apple.com/documentation/swift/uint8)) integer
value types.

In each case, the intent is to provide a clear and pragmatic syntax.

Additional discussion around this proposal's background can be found in
[#543](https://github.com/carbon-language/carbon-lang/issues/543).

## Proposal

We introduce a simple keyword-like syntax of `iN`, `uN`, and `fN` for two's
complement integers, unsigned integers, and floating-point numbers,

> respectively. Where `N` can map to common power-of-two sizes (for example,
> `N = 8, 16, 32`) or any other fixed-width value (for example, `N = 42`). This

Reviewer: ```suggestion
respectively. Where `N` can be a positive multiple of 8, including the
common power-of-two sizes (for example, `N = 8, 16, 32`). This
```

structure follows the successful precedent set by Rust and LLVM development
communities and potentially saves 40% or more on characters required compared to
other options such as `IntN` (for example, `i16` versus `Int16`).

### Non-goals

-   This does not address any considerations around the `bool` type
-   This does not provide a formal plan for the shape or mapping of the
    underlying types
    ([#767 comments](https://github.com/carbon-language/carbon-lang/issues/767#issuecomment-1214153375))

> -   This does not prescribe an official grammar for parsing these types

Reviewer: Also add a non-goal saying this proposal does not address other bit sizes, such as used in a bit field.

## Details

### Syntax

The syntax for a two's complement signed integer, the unsigned integer, and the
floating-point number corresponds to a lowercase 'i', 'u', or 'f' character,
respectively, indicating the type followed by a numeric value specifying the
width.

As a regular expression, this can be illustrated as:

```re
([iuf])([1-9][0-9]*)
```

Capture group 1 indicates either an 'i' for a two's complement signed integer
type, a 'u' for an unsigned integer type, or an 'f' for a floating-point number

> type. Capture group 2 specifies the width in bits.

Reviewer: Maybe say that the width in bits is restricted to multiples of 8 here?

`Sum` as a 32-bit two's complement signed integer.

### Syntax examples and considerations

These examples are not meant to be formal mappings. They are only used to
illustrate to the reader how the numeric type syntax may connect to the
underlying language semantics in future numeric type additions.

#### Syntax example use in common fixed-size numeric types

[As proposed by @chandlerc](https://github.com/carbon-language/carbon-lang/issues/543#issuecomment-845620894),
our heuristic for type qualification here will be "Any truly pervasive
non-aggregate/composite/compound type that is also fundamentally
size-parameterized."

These fixed-width sizes can naturally cover the common register widths of 8-,
16-, 32-, 64-, and 128-bits, as seen in other languages.

Which expands to:

-   `i8`, `i16`, `i32`, `i64`, and `i128` keyword aliases for
    `Carbon.Int(N={8,16,32,64,128})`
-   `u8`, `u16`, `u32`, `u64`, and `u128` keyword aliases for
    `Carbon.UInt(N={8,16,32,64,128})`
-   `f16`, `f32`, and `f64` keyword aliases for `Carbon.Float(N={16,32,64})`

Additionally, other type sizes may be available, depending on the platform (for

> example, `x86`), such as `f80`, `f128`, or `f256`.

Reviewer: ```suggestion
example, x86), such as `f80`, `f128`, or `f256`.
```

or

```suggestion
example, [x86](https://en.wikipedia.org/wiki/X86)), such as `f80`, `f128`, or `f256`.
```

#### Syntax example use in arbitrary fixed-size numeric types

In the more general case, this syntax can express N-width types.

-   `iN` for `Carbon.Int(N)`, An N-width two's complement signed integer type
-   `uN` for `Carbon.UInt(N)`, An N-width unsigned integer type

> -   `fN` for `Carbon.Float(N)`, An N-width IEEE 754 floating-point number type

Reviewer 3: ```suggestion
-   `fN` for `Carbon.Float(N)`, An N-width IEEE 754 binary floating-point number type
```
IEEE 754 also defines decimal floating-point formats.

an explicit goal is to provide excellent ergonomics.

Highlighting relevant aspects of this from the project goals:

-   _Carbon should not use symbols that are difficult to type, see, or
    differentiate from similar symbols in commonly used contexts._
-   _Syntax should be easily parsed and scanned by any human in any development
    environment, not just a machine or a human aided by semantic hints from an
    IDE._
-   _Explicitness must be balanced against conciseness, as verbosity and
    ceremony add cognitive overhead for the reader, while explicitness reduces
    the amount of outside context the reader must have or assume._

The type system syntax must also complement Carbon's target for
["Performance-critical software"](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/goals.md#performance-critical-software)

Specifically, there should be "No need for a lower level language."

-   _Developers should not need to leave the rules and structure of Carbon,
    whether to gain control over performance problems or to gain access to
    hardware facilities._

> ## Alternatives considered

Reviewer: Add an alternative about different sets of bit sizes that could be supported (all bit sizes, just common powers of 2).

As discussed in
[#543](https://github.com/carbon-language/carbon-lang/issues/543), three other
options were considered:

### C++ LP64 convention

Where `char` is the 8-bit type, `short` is the 16-bit type, `int` is the 32-bit
type, `long` is the 64-bit type.

Advantages:

-   The type name indicates its use to the reader
-   There is an existing precedent of this pattern in many programming
    languages, including C++
-   In the case of a typo, potentially better compiler checks versus an
    abbreviated form (for example, `i332`)

Disadvantages:

-   The type names themselves, as compared to the actual width and potentially
    use often can be arbitrary and confusing

> -   The names themselves can be longer than the other syntax options

Reviewer 3: ```suggestion
-   The names themselves can be longer than the other syntax options
-   Some common C++ implementations use other models, which may create confusion
    when interoperating with C++ code. For example, Windows uses the LLP64 model,
    where `long` is a 32-bit type, so Carbon code and C++ on Windows would have
    different and incompatible definitions for `long`.
```

floating-point number.

As these types are pervasive throughout the language, our goal here is to align
on a terse, convenient, yet understandable, and ergonomic syntax to the author.

## Background

For developer convenience, names are given to number types that map to native
machine register widths. These sizes typically include 8-bit, 16-bit, 32-bit,
64-bit, and, more recently, 128-bit widths.

For example, in [C++11+](https://en.cppreference.com/w/cpp/types/integer),
integer types such as `int8_t` (8-bit two's complement signed integer type) and
`uint16_t` (16-bit unsigned integer type) exist, among similar types for 32- and
64-bit values. Correspondingly, you have the `i8` and `u16`
([among others](https://doc.rust-lang.org/book/ch03-02-data-types.html#scalar-types))
scalar integer types in Rust. And in Swift, the `Int8` and `UInt16`
([among others](https://developer.apple.com/documentation/swift/uint8)) integer
value types.

In each case, the intent is to provide a clear and pragmatic syntax.

Additional discussion around this proposal's background can be found in
[#543](https://github.com/carbon-language/carbon-lang/issues/543).

## Proposal

> We introduce a simple keyword-like syntax of `iN`, `uN`, and `fN` for two's

Reviewer: It would be good to say somewhere that we think of these as "type literals" just like `7` is a "numeric literal".

potentially saves 40% or more on characters required compared to other options
such as `IntN` (for example, `i16` versus `Int16`). While bit sizes greater than
128-bits will be well-supported, some operations like division will not be
available on these large sizes.

### Non-goals

-   This does not address any considerations around the `bool` type
-   This does not provide a formal plan for the shape or mapping of the
    underlying types
    ([#767 comments](https://github.com/carbon-language/carbon-lang/issues/767#issuecomment-1214153375))
-   This does not prescribe an official grammar for parsing these types
-   This proposal does not address other, non-multiple of 8 bit sizes, such as
    those used in a bit field

## Details

### Syntax

The syntax for a two's complement signed integer, the unsigned integer, and the
floating-point number corresponds to a lowercase 'i', 'u', or 'f' character,
respectively, indicating the type followed by a numeric value specifying the
width.

As a regular expression, this can be illustrated as:

```re
([iuf])([1-9][0-9]*)
```

Capture group 1 indicates either an 'i' for a two's complement signed integer

> type, a 'u' for an unsigned integer type, or an 'f' for a floating-point number

Reviewer: It is good to specify that these are following the standard. It is also good to say this is a "binary" floating-point type, since that standard defines both binary and decimal floating-point formats.

```suggestion
type, a 'u' for an unsigned integer type, or an 'f' for an
[IEEE-754](https://en.wikipedia.org/wiki/IEEE_754) binary floating-point number
```

type. Capture group 2 specifies the width in bits. Note that this bit width is
restricted to a multiple of 8.

Examples of this syntax include:

-   `i16` - A 16-bit two's complement signed integer type
-   `u32` - A 32-bit unsigned integer type

> -   `f64` - A 64-bit IEEE 754 floating-point number type

Reviewer: I believe "IEEE 754" should have a hyphen if it is being used as an adjective. Here "IEEE-754", "binary", "floating-point" "number" all are modifying type:

```suggestion
-   `f64` - A 64-bit IEEE-754 binary floating-point number type
```

(Please make this change below as well.)

our heuristic for type qualification here will be "Any truly pervasive
non-aggregate/composite/compound type that is also fundamentally
size-parameterized."

These fixed-width sizes can naturally cover the common register widths of 8-,
16-, 32-, 64-, and 128-bits, as seen in other languages.

Which expands to:

-   `i8`, `i16`, `i32`, `i64`, and `i128` keyword aliases for
    `Carbon.Int(N={8,16,32,64,128})`
-   `u8`, `u16`, `u32`, `u64`, and `u128` keyword aliases for
    `Carbon.UInt(N={8,16,32,64,128})`
-   `f16`, `f32`, and `f64` keyword aliases for `Carbon.Float(N={16,32,64})`

Additionally, other type sizes may be available, depending on the platform (for
example, [x86](https://en.wikipedia.org/wiki/X86)), such as `f80`, `f128`, or
`f256`.

#### Syntax example use in arbitrary fixed-size numeric types

In the more general case, this syntax can express N-width types.

-   `iN` for `Carbon.Int(N)`, An N-width two's complement signed integer type
-   `uN` for `Carbon.UInt(N)`, An N-width unsigned integer type
-   `fN` for `Carbon.Float(N)`, An N-width IEEE 754 binary floating-point number

>     type

Reviewer: I'm not sure what these sections are adding, particularly now that there are not differences between "common" and "arbitrary" sizes.

Author: Yeah, that's fair, and I agree. I'm going to remove both of these sections.
